### **Nama : Mitslina**
### **NIM : 24060121130068**
### **1. Lakukan Eksplorasi terhadapat algoritma klasifikasi lain yang ada!**
### **Data set : Iris**

In [1]:
# import library pandas
import pandas as pd

In [2]:
# url berisi link dataset Iris
url = "http://archive.ics.uci.edu/ml/machine-learning-databases/iris/iris.data"
# memberikan nama dari fitur (kolom) yang akan digunakan
names = ['sepal-length', 'sepal-width', 'petal_length', 'petal-width','class']
# memuat dataset csv sesuai url dan names
dataset = pd.read_csv(url, names=names)

In [3]:
# melihat isi dataset
dataset.head(5)

sepal-length  sepal-width  petal_length  petal-width        class
0           5.1          3.5           1.4          0.2  Iris-setosa
1           4.9          3.0           1.4          0.2  Iris-setosa
2           4.7          3.2           1.3          0.2  Iris-setosa
3           4.6          3.1           1.5          0.2  Iris-setosa
4           5.0          3.6           1.4          0.2  Iris-setosa

In [4]:
# melihat dimensi dataset
dataset.shape

(150, 5)

### **Untuk memperkirakan akurasi pada data yang tidak diketahui sebelumnya terdapat beberapa data yang dapat dilakukan :**

### **I. Membuat Validasi Dataset**

In [5]:
# import library berikut untuk menggunakan fungsi split data training (train_test_split())
from sklearn import model_selection

In [6]:
# #persiapan data
array = dataset.values
# X menyimpan inputan fitur model machine learning yaitu fitur index-0 -- index-3
X = array[:,0:4]
# Y berisi target data, index ke 4 yaitu 'class'
Y = array[:,4]

# #pembagian data
# validasi data model 20% dari total data keseluruhan, sehingga 80% adalah data training
validation_size = 0.20
# untuk mengacak data yang akan dipakai untuk split data, utk memastikan hasil  pembagian data konsisten ketika menjalankan kode berulang kali
seed = 7
# X dan Y sesuai penjelasan yang telah disebutkan di atas, train untuk data training dan validation untuk data validasi
# model_selection.train_test_split untuk membagi dataset menjadi dua bagian yaitu data latih (training set) dan data validasi (validation set)
X_train, X_validation, Y_train, Y_validation = model_selection.train_test_split(X, Y, test_size=validation_size, random_state=seed)

### **Penjelasan:**
Berdasarkan kode tersebut kita telah membagi dataset menjadi dua bagian utama:

*   X_train dan Y_train untuk melatih model.
*   X_validation dan Y_validation untuk menguji dan mengevaluasi model.

Dengan cara ini, kita dapat menggunakan X_train dan Y_train untuk melatih model, dan kemudian menggunakan X_validation dan Y_validation untuk mengukur sejauh mana model tersebut dapat melakukan prediksi yang akurat pada data yang belum pernah dilihat sebelumnya (data validasi).

In [7]:
# melihat hasil dimensinya (record, fitur)
print("X_train :", X_train.shape)
print("X_validation :", X_validation.shape)
print("Y_train :", Y_train.shape)
print("Y_validation :", Y_validation.shape)

X_train : (120, 4)
X_validation : (30, 4)
Y_train : (120,)
Y_validation : (30,)


### **Penjelasan:**
*   80% data training : 120
*   20% data validasi : 30

In [8]:
# distribusi kelas data
dataset.groupby('class').size()

class
Iris-setosa        50
Iris-versicolor    50
Iris-virginica     50
dtype: int64

### **Penjelasan:**
Distribusi kelas data yang dihasilkan menunjukkan bahwa setiap kelas memiliki 50 sampel (distribusi seimbang karena setiap kelas memiliki jumlah sampel yang sama)

### **II. K-Folds Cross Validation**

In [9]:
# training split --> memisahkan data menjadi 2 bagian, iterasi sekali
# K fold --> memisahkan data menjadi beberapa K fold, setiap fold secara bergantian dijadikan sebagai set pengujian sedangkan yang lainnya digunakan sebagai pelatihan

# Test options and evaluation metric menggunakan accuracy (distribusi kelas seimbang)
seed = 7
scoring = 'accuracy'

### **III. Membangun Model**

In [16]:
# import library berisi algoritma klasifikasi yang akan digunakan
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import RandomForestClassifier, GradientBoostingClassifier

# Spot Check Algorithms. Evaluasi model menggunakan cross validation
models = []
models.append(('DT', DecisionTreeClassifier()))
models.append(('RF', RandomForestClassifier()))
models.append(('GB', GradientBoostingClassifier()))

### **Penjelasan:**
Pada kode tersebut, akan mengevaluasi dengan beberapa algoritma untuk mengetahui algoritma yang cocok dengan studi kasus ini. Model yang digunakan yaitu decision tree (DT), random forest (RF), dan gradient boosting (GB)

Kita akan menggunakan validasi silang 10 kali lipat untuk memperkirakan akurasi. Untuk itu dataset dibagi menjadi 10 bagian, 9 untuk latihan dan 1 untuk pengujian dan ulangi untuk semua kombinasi.

In [17]:
# evaluate each model in turn

# menyimpan hasil evaluasi K fold validation dari masing-masing model
results = []
# menyimpan nama model
names = []

for name, model in models:
  # membagi data menjadi training dan validation menggunakan kfold validation, dengan k atau n_splits nya adalah 10 dan random_state = 7
  # mengaktifkan shuffle=True. shuffle digunakan untuk menentukan apakah data harus diacak atau tidak sebelum dibagi
  kfold = model_selection.KFold(n_splits=10, random_state=seed, shuffle=True)
  # model_selection.cross_val_score untuk menghitung nilai scoring dari setiap iterasi
  # mengukur kinerja dari data dari model yang sudah didefinisikan dengan data latih yang sudah di split (X_train dan Y_train), cv = kfold yg sudah didefinisikan, scoring = accuracy
  cv_results = model_selection.cross_val_score(model, X_train, Y_train, cv=kfold, scoring=scoring)
  # memasukan cv_results ke dalam result untuk setiap iterasi
  results.append(cv_results)
  # memasukan name ke dalam names untuk setiap iterasi
  names.append(name)
  # menampilkan pesan nama model, rata2 final, dan stdev final
  msg = "%s: %f (%f)" % (name, cv_results.mean(), cv_results.std())
  print(msg)

DT: 0.950000 (0.076376)
RF: 0.966667 (0.040825)
GB: 0.950000 (0.055277)


### **Penjelasan:**
*   Model Desicion Tree menghasilkan final rata-rata=0.950000 dan stdev=0.076376
*   Model Random Forest menghasilkan final rata-rata=0.966667  dan stdev=0.040825
*   Model Gradient Boosting menghasilkan final rata-rata=0.950000  dan stdev=0.055277

### **IV. Memilih Model Terbaik**

### **Penjelasan:**
Berdasarkan hasil output tersebut di atas, kita dapat melihat bahwa Random Forest (RD) memiliki nilai akurasi perkiraan terbesar. Setelah mengetahui model yang paling akurat yaitu RD, selanjutnya kita dapat mencoba melakukan pengujian tentang keakuratan model RD terhadap data yang ada.

In [18]:
from sklearn.metrics import accuracy_score, confusion_matrix, classification_report

In [25]:
# Membuat prediksi validasi dataset

RF = RandomForestClassifier()
# melatih (fit) model RF menggunakan data latih
RF.fit(X_train, Y_train)
# menguji data RF yang dihasilkan dari proses pelatihan data dengan testing data
predictions = RF.predict(X_validation)

# menampilkan hasil
print('accuracy score: ', accuracy_score(Y_validation, predictions))
print('confusion matrix: \n', confusion_matrix(Y_validation, predictions))
print('\n\t\t classification report: \n', classification_report(Y_validation, predictions))

accuracy score:  0.8666666666666667
confusion matrix: 
 [[ 7  0  0]
 [ 0 10  2]
 [ 0  2  9]]

		 classification report: 
                  precision    recall  f1-score   support

    Iris-setosa       1.00      1.00      1.00         7
Iris-versicolor       0.83      0.83      0.83        12
 Iris-virginica       0.82      0.82      0.82        11

       accuracy                           0.87        30
      macro avg       0.88      0.88      0.88        30
   weighted avg       0.87      0.87      0.87        30



### **Penjelasan:**
Output yang dihasilkan dari evaluasi model memberikan gambaran tentang seberapa baik model Random Forest (RF) dapat melakukan klasifikasi pada data validasi.

- Accuracy Score:
Akurasi adalah rasio prediksi yang benar (positif dan negatif) terhadap total data. Dalam konteks ini, akurasi sebesar 0.867 (86.7%) menunjukkan bahwa model RF dapat mengklasifikasikan data dengan benar sekitar 86.7% pada saat data validasi.

- Confusion Matrix:
Matriks konfusi adalah tabel yang menunjukkan jumlah prediksi yang benar dan salah untuk setiap kelas. Dalam kasus ini, matriks konfusi memiliki tiga baris dan tiga kolom, yang sesuai dengan tiga kelas Iris (setosa, versicolor, virginica).
 - Baris pertama menunjukkan kelas Iris-setosa: Ada 7 prediksi benar (True Positive), dan tidak ada prediksi salah untuk kelas ini.
 - Baris kedua menunjukkan kelas Iris-versicolor: Ada 10 prediksi benar, 2 prediksi salah sebagai Iris-virginica.
 - Baris ketiga menunjukkan kelas Iris-virginica: Ada 9 prediksi benar, 2 prediksi salah sebagai Iris-versicolor.

- Classification Report:
Laporan klasifikasi memberikan matriks precision,  recall, f1-score untuk setiap kelas, serta rata-rata dari matriks tersebut untuk semua kelas (macro avg dan weighted avg).
 - Precision: Seberapa akurat model dalam mengklasifikasikan positif. Misalnya, presisi untuk Iris-setosa adalah 1.00, yang berarti setiap prediksi Iris-setosa benar.
 - Recall: Seberapa baik model dapat menangkap semua instance dari suatu kelas. Misalnya, recall untuk Iris-versicolor adalah 0.83, yang berarti model berhasil menangkap 83% dari semua instance Iris-versicolor.
 - F1-score: Rata-rata dari precision dan recall.
 - Support: Jumlah instance yang sebenarnya dari setiap kelas.

Berdasarkan hasil yang diberikan, model memiliki kinerja baik pada kelas Iris-setosa dengan presisi, recall, dan F1-score sebesar 1.00. Untuk kelas Iris-versicolor dan Iris-virginica, model memiliki presisi, recall, dan F1-score di sekitar 0.82-0.83.